New generic JDBC Persistence Service

lewie opened the pull request:
[Her you go to pre Release DBMS](https://github.com/openhab/openhab/issues/3147)
DBMS is renamed to JDBC and is is much smaller, faster, cleaner code and easier to extend than the first attempt.

lewie commented:
### New JDBC Persistence Service 

#### Keys:
- Clearly separated Source for the database-specific part of OpenHab logic.
- Code duplication by similar services can be prevented
- To integrate a new (SQL/JDBC enabled) database is fairly simple.
- Tablenames are more readable
- Tablenames can be reconfigured
- Can work with original Databasemodel of MySQL-, H2-, PostgreSql-Service
- Currently supports the following databases: derby, h2, hsqldb, mariadb, mysql, postgresql, Sqlite
- The JDBC drivers are not compiled with into the bundle. Simply copy Driverfile and Bundle into OpenHab addons. Driver files do not increase the main JDBC Service Bundle.
- It is fast and scalable

#### What is the benefit of a 'per Item' table layout?
What is the benefit to manage 500,000 Integer, Numbers or Strings per item, instead of 10,000,000 Strings in one table? An attempt to explain:
 1. Data Mass searching: 500,000 to 10,000,000.
 2. It is expensive to search for average, min, max, etc., especially in a bunch of strings.
 3. There are Items which generates only some Data and other millions of datasets per Year, and others are often queried per Minute. On Item level, these different requirements are easier to handle.
 4. Clarity
 5. Delete an Items Table and start over in a second.

With a 'per Item' layout, the scalability and easy maintenance of the database is ensured, even if large amounts of data must be managed.

#### For Testing:
[Pre-Build] (https://github.com/lewie/openhab/tree/jdbc-new-service/bundles/persistence/org.openhab.persistence.jdbc/build)
[Drivers](https://github.com/lewie/openhab/tree/jdbc-new-service/bundles/persistence/org.openhab.persistence.jdbc/drivers)


#### Configuration:
```
############################ JDBC Persistence Service ##################################
# I N S T A L L   J D B C   P E R S I S T E N C E   S E R V I C E 
# To use this JDBC-service-bundle (org.openhab.persistence.jdbc-X.X.X.jar),
# a appropriate JDBC database-driver is needed in OpenHab addons Folder.
# Copy both (JDBC-service-bundle and a JDBC database-driver) to your OpenHab '[OpenHab]/addons' Folder to make it work. 
#
# Driver jars:
# Derby:     derby-10.11.1.1.jar               http://mvnrepository.com/artifact/org.apache.derby/derby
# H2:        h2-1.4.189.jar                    http://mvnrepository.com/artifact/com.h2database/h2
# HSQLDB:    hsqldb-2.3.3.jar                  http://mvnrepository.com/artifact/org.hsqldb/hsqldb
# MariaDB:   mariadb-java-client-1.2.0.jar     http://mvnrepository.com/artifact/org.mariadb.jdbc/mariadb-java-client
# MySQL      mysql-connector-java-5.1.36.jar   http://mvnrepository.com/artifact/mysql/mysql-connector-java
# PostgreSQL:postgresql-9.4-1201-jdbc41.jar    http://mvnrepository.com/artifact/org.postgresql/postgresql
# SQLite:    sqlite-jdbc-3.8.11.1.jar          http://mvnrepository.com/artifact/org.xerial/sqlite-jdbc
#
# Tested databases/url-prefix: jdbc:derby, jdbc:h2, jdbc:hsqldb, jdbc:mariadb, jdbc:mysql, jdbc:postgresql, jdbc:sqlite
# 
# derby, h2, hsqldb, sqlite can be embedded, 
# If no database is available, for Example the url 'jdbc:h2:./testH2' creates a new DB in OpenHab Folder. 

# D A T A B A S E  C O N F I G
# Some URL-Examples, 'service' identifies and activates internally the correct jdbc driver.
# required database url like 'jdbc:<service>:<host>[:<port>;<attributes>]'
# jdbc:url=jdbc:derby:./testDerby;create=true
# jdbc:url=jdbc:h2:./testH2
# jdbc:url=jdbc:hsqldb:./testHsqlDb
# jdbc:url=jdbc:mariadb://192.168.0.1:3306/testMariadb
# jdbc:url=jdbc:mysql://192.168.0.1:3306/testMysql
# jdbc:url=jdbc:postgresql://192.168.0.1:5432/testPostgresql
# jdbc:url=jdbc:sqlite:./testSqlite.db

# required database user
#jdbc:user=
jdbc:user=test

# required database password
#jdbc:password=
jdbc:password=test

# E R R O R   H A N D L I N G
# optional when Service is deactivated (optional, default: 0 -> ignore) 
#jdbc:errReconnectThreshold=

# I T E M   O P E R A T I O N S
# optional tweaking SQL datatypes
# see: https://mybatis.github.io/mybatis-3/apidocs/reference/org/apache/ibatis/type/JdbcType.html	
# see: http://www.h2database.com/html/datatypes.html
# see: http://www.postgresql.org/docs/9.3/static/datatype.html
# defaults:
#jdbc:sqltype.CALL			=	VARCHAR(200)
#jdbc:sqltype.COLOR			=	VARCHAR(70)
#jdbc:sqltype.CONTACT		=	VARCHAR(6)
#jdbc:sqltype.DATETIME		=	DATETIME
#jdbc:sqltype.DIMMER		=	TINYINT
#jdbc:sqltype.LOCATION		=	VARCHAR(30)
#jdbc:sqltype.NUMBER		=	DOUBLE
#jdbc:sqltype.ROLLERSHUTTER	=	TINYINT
#jdbc:sqltype.STRING		=	VARCHAR(65500)
#jdbc:sqltype.SWITCH		=	VARCHAR(6)

# For Itemtype "Number" default decimal digit count (optional, default: 3) 
#jdbc:numberDecimalcount=

# T A B L E   O P E R A T I O N S
# Tablename Prefix generation, using Item real names or "item" (optional, default: false -> "item") 
#jdbc:tableUseRealItemNames=
jdbc:tableUseRealItemNames=true

# Tablename Suffix length (optional, default: 4 -> 0001-9999) 
#jdbc:tableIdDigitCount=

# Rename existing Tables using tableUseRealItemNames and tableIdDigitCount (optional, default: false) 
# USE WITH CARE! Deactivate after Renaming is done!
#jdbc:rebuildTableNames=true

# D A T A B A S E  C O N N E C T I O N S
# Some embeded Databases can handle only one Connection (optional, default: configured per database in packet org.openhab.persistence.jdbc.db.* )
# see: https://github.com/brettwooldridge/HikariCP/issues/256
# jdbc.maximumPoolSize = 1
# jdbc.minimumIdle = 1

# T I M E K E E P I N G
# (optional, default: false) 
#jdbc:enableLogTime=true

```


#### Time for some not representative Tests in my environment, with a script like this:
````
var count = 0;
rule "DB STRESS TEST"
when 
	Time cron "30 * * * * ?"
then
	count = count+1
	if( count > 3 && count < 23){
		for( var i=500; i>1; i=i-1){
			postUpdate( NUMBERITEM, i)
			SWITCHITEM.previousState().state
			postUpdate( DIMMERITEM, OFF)
			NUMBERITEM.changedSince( now().minusMinutes(1))
			postUpdate( DIMMERITEM, ON)
		}
	}
end
````
Each Test ran about 20 Times every 30 seconds.
Openhab has ready started for about a Minute.
The data in seconds for the evaluation I had from the Console Output.

#### Results per DATABASE:
DATABASE | FIRST RUN | AVERAGE | FASTEST | COMMENT
------------- | ------------- | ------------- | ------------- | -------------
Derby | 7.829 | 6.892 | 5.381 | local embedded
H2 | 1.797 | 2.080 | 1.580 | local embedded
hsqldb | 3.474 | 2.104 | 1.310 | local embedded
mysql | 11.873 | 11.524 | 10.971 | external Server VM
postgresql | 8.147 | 7.072 | 6.895 | external Server VM
sqlite | 2.406 | 1.249 | 1.137 | local embedded

##### DATABASE file sizes after these Tests:
Derby  = 5.36 MB
H2     = 0.96 MB
hsqldb = 1.23 MB
sqlite = 0.28 MB
	
#### Results per OpenHab-Bundle:	
OH-Bundle | DATABASE | FIRST RUN | AVERAGE | FASTEST | COMMENT
------------- | ------------- | ------------- | ------------- | ------------- | -------------
JDBC | mysql | 11.873 | 11.524 | 10.971 | external Server VM
DBMS | mysql | 17.375 | 15.867 | 15.341 | external Server VM
JPA | mysql | 20.776 | 13.872 | 12.761 | external Server VM
MYSQL | mysql | 13.751 | 12.867 | 12.490 | external Server VM



Thanks and Have Fun!


Rgds
Helmut

grzech1983 commented:
Hello,

one more thing. Are you testing your persistance addon also on OH2? I'm asking because I would like to report the same thing I had with DBMS addon. I've set JDBC in every place I could and I'm still getting:
```
HTTP ERROR 500

Problem accessing /chart. Reason:

    Server Error
Caused by:

java.lang.NoSuchMethodError: java.sql.Timestamp.toInstant()Ljava/time/Instant;
	at org.openhab.persistence.jdbc.db.JdbcBaseDAO.objectAsDate(JdbcBaseDAO.java:435)
	at org.openhab.persistence.jdbc.db.JdbcBaseDAO.doGetHistItemFilterQuery(JdbcBaseDAO.java:294)
	at org.openhab.persistence.jdbc.internal.JdbcMapper.getHistItemFilterQuery(JdbcMapper.java:137)
	at org.openhab.persistence.jdbc.internal.JdbcPersistenceService.query(JdbcPersistenceService.java:294)
	at org.openhab.core.persistence.internal.QueryablePersistenceServiceDelegate.query(QueryablePersistenceServiceDelegate.java:49)
	at org.eclipse.smarthome.ui.internal.chart.DefaultChartProvider.addItem(DefaultChartProvider.java:251)
	at org.eclipse.smarthome.ui.internal.chart.DefaultChartProvider.createChart(DefaultChartProvider.java:171)
	at org.eclipse.smarthome.ui.internal.chart.ChartServlet.doGet(ChartServlet.java:171)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.equinox.http.servlet.internal.ServletRegistration.service(ServletRegistration.java:61)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.processAlias(ProxyServlet.java:128)
	at org.eclipse.equinox.http.servlet.internal.ProxyServlet.service(ProxyServlet.java:60)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:808)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:215)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:110)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:310)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)

```

JDBC log shows:

```
2015-10-05 07:57:53.880 DEBUG o.o.p.j.i.JdbcPersistenceService[:257]- JDBC::query: item is HEM_E
2015-10-05 07:57:53.882 DEBUG o.o.p.jdbc.internal.JdbcMapper[:134]- JDBC::getHistItemFilterQuery filter='true' numberDecimalcount='3' table='item0010' item='HEM_E (Type=NumberItem, State=1372.56)' itemName='HEM_E'
2015-10-05 07:57:53.883 DEBUG o.o.p.jdbc.db.JdbcBaseDAO[:303]- JDBC::getHistItemFilterQueryProvider filter = org.openhab.core.persistence.FilterCriteria@47594e4f, numberDecimalcount = 3, table = item0010, simpleName = HEM_E
2015-10-05 07:57:53.884 DEBUG o.o.p.jdbc.db.JdbcBaseDAO[:323]- JDBC::query queryString = SELECT time, value FROM item0010 WHERE TIME>'2015-10-04 07:57:53' AND TIME<'2015-10-05 07:57:53' ORDER BY time ASC
2015-10-05 07:57:53.885 DEBUG o.o.p.jdbc.db.JdbcBaseDAO[:289]- JDBC::doGetHistItemFilterQuery sql=SELECT time, value FROM item0010 WHERE TIME>'2015-10-04 07:57:53' AND TIME<'2015-10-05 07:57:53' ORDER BY time ASC
2015-10-05 07:57:54.159 DEBUG o.o.p.jdbc.db.JdbcBaseDAO[:397]- JDBC::ItemResultHandler::handleResult getState value = '3592.08', getClass = 'class java.lang.Double', clazz = 'Double'

```

so as you can see item is being queries correctly. Any ideas?

lewie commented:
@grzech1983,

this has nothing to do with OpenHab 1 or 2.
I accidentally have used new JAVA 8 ````java.sql.Timestamp.toInstant();```` function.
Your OpenHab 2 Environment runs on JAVA 7 (highest version for OpenHab 1), so you ran into this Bug.

It is fixed now.

I tested only with JAVA 8, while using javascript for Testing the Service.

This bundle is still very new, please do not be angry if I do have overlooked something. ;-)

Thank You
Helmut

grzech1983 commented:
Hello,

I can confirm it works. I'm faaar from being angry. I wanted to be helpfull here and I'm glad that you've found what was causing this problem.

Thanks

lewie commented:
Thank you, you're dehr helpful!
Critical testers need to be able to harden software.

Rgds
Helmut

grzech1983 commented:
After few days of using updated jdbc addon I don't see any issues. It work as expected (I'm using MySQL database).

watou commented:
Out of curiosity, is the "default" installation of MySQL support compatible with the database and tables created by the existing MySQL persistence bundle?  If not, should it be?

grzech1983 commented:
IMHO it is not but now I don't remembed what was previous DB schema (I've removed it during testing). Here you have items table create statement:

```
SELECT * FROM openhab.items;CREATE TABLE `items` (
  `ItemId` int(11) NOT NULL AUTO_INCREMENT,
  `itemname` varchar(500) NOT NULL,
  PRIMARY KEY (`ItemId`)
) ENGINE=InnoDB AUTO_INCREMENT=37 DEFAULT CHARSET=utf8;
```
and example item:

```
CREATE TABLE `item0001` (
  `time` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `value` double DEFAULT NULL,
  PRIMARY KEY (`time`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

lewie commented:
@watou, yes, tables created by the existing MySQL persistence bundle are compatible with JDBC persistence bundle. If not it is a Bug!
Note, ````tableUseRealItemNames```` must not have value 'true' (Tablenames are like 'items0001'). 

@grzech1983, To create a MySQL database for OH use:
````CREATE DATABASE 'yourDB' CHARACTER SET utf8 COLLATE utf8_general_ci;````

Rgds
Helmut

watou commented:
Thanks @grzech1983!
@lewie, wouldn't a setting named `tableUseRealItemNames` need to be set to `false` in order to use the same numbering scheme as the MySQL persistence bundle?  Also, in my instance of the MySQL persistence bundle, my table names are Items, Item1, Item2, ..., Item93.  Note the case difference in the table name and also the absence of leading zeros.  Is this really compatible with the new JDBC persistence bundle?

lewie commented:
@watou, You are right!
Updated Description, code and pre-build. 
Now you can set 'tableNamePrefix' in configuration.
````
# Tablename Prefix String (optional, default: "item") 
# for Migration from MYSQL-Bundle set to 'Item'.
jdbc:tableNamePrefix=Item
````

lewie commented:
Found a real bad Problem,

Some time ago, I had very long decimals when data comes from 'double' in databases.
Therefore, I then introduced in my PostgreSQL bundle the parameter 'numberDecimalcount'.
Rounding aka 'numberDecimalcount' is off at the moment, so decimals are back... 


Using double for Numbers causes (as in Mysql bundle) rounding trouble without end.
For Example 1.005 can not be represented in binary floating point double in java.
Have a look at this discussion: http://stackoverflow.com/questions/153724/how-to-round-a-number-to-n-decimal-places-in-java

It seams the only right way, OpenHab 'DecimalType' not to save it as a 'DOUBLE' but as a jdbc-type 'NUMERIC' and 'DECIMAL'. The Java equivalent is 'BigDecimal'. OpenHab DecimalType holds its data as BigDecimal anyway.

 1. We would loose backwards compatibility to MYSQL bundel or it has to be programmed a converter/switch.
 2. It is never as fast as now.

Any suggestions?
I think we want to do it right?

Rgds
Helmut

watou commented:
Thanks @lewie!  Why does `jdbc:tableIdDigitCount` default to 4?  Why have all those extra zeroes in the table names?  If it's arbitrary, I would rather the default is 0, so compatibility with MySQL persistence doesn't have to be labour-intensive, instead a "just works" replacement.  Also, why not have `jdbc:tableNamePrefix` be the same as the MySQL bundle's choice, again to avoid special configuration?

I worry that there are so many configuration options, where instead there would be instant compatibility with the MySQL persistence bundle with no extra steps, and sensible choices, where they need to differ from MySQL, would be made in your per-DB classes.

About the issue with DOUBLE vs. NUMERIC etc., possibly @cdjackson remembers why DOUBLE was chosen in the MySQL persistence bundle?

lewie commented:
Hi @watou,

tableIdDigitCount: item0001 - item9999 easier to read, sort and parse as item1 - item9999.
Sorting item1 - item9999 unusable:
item1
item10
item11
...
item1111
item2
item20
item21
...

tableNamePrefix: Is default 'item' because it makes sense to use Upper- OR lowercase in Tablenames. To use upper- and lowercase can cause Problems with other Databases. Some Databases do convert all Letters uppercase, so in JDBC bundle it makes sense to use either or, to keep it simpler and all Databases easily maintainable together in one bundle.

How often is Bundle switching used? Is a special case, it is only needed in a transitional period.
Later People use MYSQL- or JDBC-bundle and then stick to it.
Instant compatibility with the MySQL persistence bundle, it is a Feature, not the main task.

Rgds
Helmut

watou commented:
Thanks for the reply and explanation, @lewie.  In that case, it would be good to have a small section in the [wiki page](https://github.com/openhab/openhab/wiki/JDBC-Persistence) that concisely documents how to configure the JDBC persistence bundle to compatibly pick up where the [MySQL persistence bundle](https://github.com/openhab/openhab/wiki/MySQL-Persistence) left off.  Are you planning on making a wiki page for the JDBC persistence bundle, or is it there under a different name already?

lewie commented:
@watou, 
yes, you're right, have to describe it clean naturally.
I collect and pre-format the information currently [here at the top of the post.](https://github.com/openhab/openhab/pull/3248#issuecomment-145394680)

This information I use directly on the wiki page later.

In the configuration instructions are hints.
In section ['Migrate from MYSQL to bundle JDBC bundle:'](https://github.com/openhab/openhab/pull/3248#issuecomment-145394680) is a small section.

Thank You
Helmut

lewie commented:
Thanks @watou,

Now I have filled the [wiki page](https://github.com/openhab/openhab/wiki/JDBC-Persistence).
Any suggestions for content and typos?!

Rgds
Helmut

watou commented:
I made edits to [your wiki page](https://github.com/openhab/openhab/wiki/JDBC-Persistence).

A remaining area that lacks clarity are the specific steps to have the bundle perform a conversion from numbered to named tables.  If that were laid out clearly in the wiki, from start to finish, it would stop the inevitable questions that would follow from confused users.

lewie commented:
Well done!
I can not see why, but Anchor links for 'Installation' and 'Configuration' do not jump.

watou commented:
Thanks for catching that.  I changed the section headings and anchor links and now it works.  Mystery.

lewie commented:
@watou, Thanks for perfect tweaking [wiki page](https://github.com/openhab/openhab/wiki/JDBC-Persistence), Thank You. :-)


Rgds
Helmut

watou commented:
You're very welcome, Helmut.  Thank you for all your work in contributing this great new persistence package.

lewie commented:
Added two Points to [wiki page](https://github.com/openhab/openhab/wiki/JDBC-Persistence):
Number Precision and Rounding results

lewie commented:
Hello @teichsta, 
if you find time, `New generic JDBC Persistence Service` seems to be ready to be admin reviewed.

Rgds
Helmut

hakan42 commented:
Hi @lewie , could you please rebase this against the current master? I would very much love to get this into 1.8.0 :smile:

lewie commented:
Hi @hakan42, re-based it. 
Very good that Thomas gets your support! 

Thanks for review my JDBC Persistence Service. :-)

Info: 
Tested today in Openhab 2, works well.

Rgds
Helmut

hakan42 commented:
@lewie : The code is not formatted according to the OH2 eclipse styles yet, I will go forward with the review. After all the review is finished, we can reformat all the code at once and squash the commit into one.

I'll do some further comments later today and write a note here when I am finished with the first round.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/internal/JdbcConfiguration.java:
```diff
@@ -0,0 +1,321 @@
+			logger.error(
+					"JDBC::updateConfig: SQL password is missing. Attempting to connect without password. To specify a password configure the jdbc:password parameter in openhab.cfg.");
+		}else{
+			logger.debug("JDBC::updateConfig:  password={}{}...", password.subSequence(0, 3), password.length() - 3);
```
nitpick: IndexOutOfBoundsException would happen if the password is shorter than 4 characters.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/internal/JdbcPersistenceService.java:
```diff
@@ -0,0 +1,333 @@
+	 * set in the activate() method and must not be accessed anymore once the
+	 * deactivate() method was called or before activate() was called.
+	 */
+	@SuppressWarnings("unused")
```
Is this SuppressWarnings necessary? The bundleContext is being used in the activate() method...

lewie replied:
Yes it is needed, Eclipse says... ;-)

lewie commented:
@hakan42 , 
oh, where can I find the OH2 eclipse style Guide?

hakan42 commented:
The complete OH2 development environment ( https://community.openhab.org/t/setting-up-a-development-environment/2445 ) contains the code formatter configuration too.

As this is a completely new binding, I would love if we start off with already the "correct" code style :smile:

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/lib/.gitignore:
```diff
@@ -0,0 +1,3 @@
+/commons-dbutils-1.6.jar
```
please put wildcards like "commons-dbutils.*" in this gitignore so this file does not need to be updated with each library update

hakan42 commented:
While not strictly required for OH1, could you add an about.html similiar to https://github.com/openhab/openhab2/blob/master/addons/binding/org.openhab.binding.avmfritz/about.html showing the additional libraries you are using and the licenses they are distributed with?

watou commented:
Is it correct to use the OH2 code style in the OH1 repo?

hakan42 commented:
Whatever style we use, it should be automatically applied. As the OH2 configuration is the only one I know of, I suggested that one :smile:

watou commented:
I recommend that all code in the same repo uses the same code formatting.

hakan42 commented:
No matter which style we go with, everything is better than manually formating code :smiling_imp: 

The thing is, all I can find about OH1 style is the sentence "Code must be formatted using the provided code formatter and clean up settings (import them into your IDE)." from the OH1 how-to-contribute wiki page ( https://github.com/openhab/openhab/wiki/How-To-Contribute ).

@watou , do you know where the settings can be found?

lewie commented:
@hakan42, 
yes we should start off with already the "correct" code style!!! 
Do you simply mean the Source Formatter of Eclipse? Ctrl+Shift+F

Exactly there I search too now... "Code must be formatted using the provided code formatter and clean up settings (import them into your IDE)."

Hmm, "import them into your IDE", yes really very happy, but from where? 

Can't find it here: https://github.com/openhab/openhab2/blob/master/docs/sources/development/ide.md
or anywhere else. ;-)

Sorry
Helmut

watou commented:
It comes as part of the Yoxos install for the OH1 IDE setup.  One of the PRs I reviewed recently went into more detail (on mobile now so hard to look up).

lewie commented:
Thank you @watou, do not torture! :-)
I simply install the [OH2 IDE](https://github.com/openhab/openhab2/blob/master/docs/sources/development/ide.md) and export them to me.

watou commented:
Some comments about formatting [here](https://github.com/openhab/openhab/pull/3187#issuecomment-152722012)

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/pom.xml:
```diff
@@ -0,0 +1,290 @@
+		<hikari.version>2.4.1</hikari.version>
+		<!-- <hikari.version>2.3.9</hikari.version> <hikari.version>2.4.1</hikari.version> -->
+		<dbutils.version>1.6</dbutils.version>
+		<yank.version>3.0.3-SNAPSHOT</yank.version>
```
If we use yank 3.0.3-SNAPSHOT, please add it to lib (with appropriate entry in about.html). If we can stay with 3.0.2 (the latest version in maven central), please set this in pom.xml and MANIFEST.MF

lewie replied:
We will get a [new yank release 3.1.0](https://github.com/timmolter/Yank/issues/27#issuecomment-159532252) on maven central by timmolter.

lewie commented:
@hakan42, are you ready with first review round?

hakan42 commented:
sorry, no, I was actually waiting for the rebase, and then got bogged down in work. I'll go through the code in the next hour or so, and write a notice here as soon as I am done.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/db/JdbcBaseDAO.java:
```diff
@@ -0,0 +1,530 @@
+	private String histItemFilterQueryProvider(FilterCriteria filter, int numberDecimalcount, String table, String simpleName) {
+		logger.debug("JDBC::getHistItemFilterQueryProvider filter = {}, numberDecimalcount = {}, table = {}, simpleName = {}", filter.toString(), numberDecimalcount, table, simpleName);
+		
+		SimpleDateFormat jdbcDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
This DateFormat object never changes. Maybe make it static final in the class scope and save a bit of performance?

watou replied:
That class isn't threadsafe, so making it static could lead to mysterious issues.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/db/JdbcDerbyDAO.java:
```diff
@@ -0,0 +1,211 @@
+	private String histItemFilterQueryProvider(FilterCriteria filter, int numberDecimalcount, String table, String simpleName) {
+		logger.debug("JDBC::getHistItemFilterQueryProvider filter = {}, numberDecimalcount = {}, table = {}, simpleName = {}", filter.toString(), numberDecimalcount, table, simpleName);
+		
+		SimpleDateFormat jdbcDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
This DateFormat object never changes. Maybe make it static final in the class scope and save a bit of performance?

watou replied:
Ibid.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/db/JdbcPostgresqlDAO.java:
```diff
@@ -0,0 +1,210 @@
+		sqlTypes.put("ROLLERSHUTTERITEM", "SMALLINT");
+		sqlTypes.put("STRINGITEM", "VARCHAR");
+		sqlTypes.put("SWITCHITEM", "VARCHAR");
+		logger.debug("JDBC::initPostgreSqlTypes: Initialize the type array sqlTypes={}", sqlTypes.values());
```
minor nitpick: logger message is not equal to the method name :smile:

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/db/JdbcPostgresqlDAO.java:
```diff
@@ -0,0 +1,210 @@
+	private String histItemFilterQueryProvider(FilterCriteria filter, int numberDecimalcount, String table, String simpleName) {
+		logger.debug("JDBC::getHistItemFilterQueryProvider filter = {}, numberDecimalcount = {}, table = {}, simpleName = {}", filter.toString(), numberDecimalcount, table, simpleName);
+		
+		SimpleDateFormat jdbcDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
```
This DateFormat object never changes. Maybe make it static final in the class scope and save a bit of performance?

watou replied:
Ibid.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/internal/JdbcConfiguration.java:
```diff
@@ -0,0 +1,321 @@
+		// Which DB-Type to use
+		serviceName = parsedURL.getProperty("dbShortcut"); // derby, h2, hsqldb, mariadb, mysql, postgresql, sqlite
+		logger.debug("JDBC::updateConfig: found serviceName = '{}'", serviceName);
+		if (StringUtils.isBlank(serviceName)) {
```
minor nitpick: make sure that serviceName is at least two characters, else you will run into IndexOutOfBoundsException a few lines later.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/internal/JdbcConfiguration.java:
```diff
@@ -0,0 +1,321 @@
+		else if(serviceName.equals("mysql"))      warn += "\tMySQL      Tested version: mysql-connector-java-5.1.36.jar   http://mvnrepository.com/artifact/mysql/mysql-connector-java\n";
+		else if(serviceName.equals("postgresql")) warn += "\tPostgreSQL:Tested version: postgresql-9.4-1201-jdbc41.jar    http://mvnrepository.com/artifact/org.postgresql/postgresql\n";
+		else if(serviceName.equals("sqlite"))     warn += "\tSQLite:    Tested version: sqlite-jdbc-3.8.11.1.jar          http://mvnrepository.com/artifact/org.xerial/sqlite-jdbc\n";
+		
```
Maybe put the exact tested library versions on the wiki and just write "check wiki for versions" here? Else, you would have to remember to update those warning messages every time you update a library.

hakan42 replied:
Also, where is this "warn" used? I don't see a logging statement that writes this...

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/internal/JdbcPersistenceService.java:
```diff
@@ -0,0 +1,333 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
```
After the reviews are finished, please move this comment block to either the config file or the wiki

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/java/org/openhab/persistence/jdbc/utils/StringUtilsExt.java:
```diff
@@ -0,0 +1,265 @@
+        return arr;
+    }
+    
+    public static void main(String[] args) throws Exception {
```
Is this main() method for testing purposes? In that case, could it be moved to an unit test? I believe we have junit available.

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/pom.xml:
```diff
@@ -0,0 +1,290 @@
+				</configuration>
+			</plugin>
+
+<!-- http://stackoverflow.com/questions/8726884/create-multiple-runnable-jars-with-depencies-included-from-a-single-maven-proj -->
```
Please remove those commented out sections in the pom

hakan42 commented:
Sorry, I did not see that you were using java.text.SimpleDateFormat. In that case, replace SimpleDateFormat with joda DateTimeFormatter :smile: 

http://stackoverflow.com/questions/21136858/how-to-change-the-simpledateformat-to-jodatime

hakan42 commented:
You have many "magic" string constants in your initialization class. Can those be moved to a Constants class? Also, does the Yank library have constant string definitions for their configuration?

hakan42 commented:
@lewie o.k., I'm through with the review comments. After processing them, could you please run the OH code formatter over the complete source so we have it in the "one true style"?

Many thanks in advance :smile:

mdbergmann commented:
You put quite a bit of work into this.

One question: how does it work that classloader loads any driver jars from addons folder?

lewie commented:
@mdbergmann,
nearly all JDBC drivers are OSGIfied (look into driver.jar if there is a MANIFEST.MF File).
if you put any OSGI Bundle into addons folder, openHAB will load this bundle (aka driver.jar)

Thats it.

It took me a while to understand this... :-)

lewie commented:
@hakan42, 
I think it is ready to merge. 

- DONE: nitpick: IndexOutOfBoundsException would happen if the password is shorter than 4 characters.
- Needed: Is this SuppressWarnings necessary?
- DONE: please put wildcards like "commons-dbutils.*" in this gitignore so this file does not need to be updated with each library update
- DONE: If we use yank 3.0.3-SNAPSHOT... -> we get a new Version 3.1.0 in Maven from timmolter/Yank
- REMOVED: This DateFormat object never changes... -> SimpleDateFormat replaced by joda
- DONE: minor nitpick: logger message...
- DONE: minor nitpick: make sure that serviceName is at least two characters
- REMOVED: Maybe put the exact tested library versions...
- DONE: Also, where is this "warn" used?
- REMOVED: After the reviews are finished, please move this comment block...
- REMOVED: Is this main() method
- REMOVED: Please remove those commented out sections in the pom
- DONE: run the OH code formatter over the complete source


- hakan42:"You have many "magic" string constants in your initialization class. Can those be moved to a Constants class? Also, does the Yank library have constant string definitions for their configuration?" ---> This is the only point which is not so clear. Did you mean in JdbcConfiguration? This I have not done yet, because the Structure is like in other Services. Is this OK for you?


I very Thank you for your review! :-)

Helmut

hakan42 commented:
Thank you for your efforts. Tomorrow I will check the new state and then I think we can merge this :smile:

lewie commented:
any update, @hakan42?
Would be great to see this merged.
Then I could cleanup my eclipse branch chaos. ;-)

Thanks, 
Helmut

hakan42 commented:
I cannot promise an update today, but on Sunday I will have another look and hopefully merge things. Thank you for your patience :smile:

lewie commented:
Thank You, all right! :-)

hakan42 commented on bundles/persistence/org.openhab.persistence.jdbc/about.html:
```diff
@@ -0,0 +1,38 @@
+    for informational purposes only, and you should look to the Redistributor's license for
+    terms and conditions of use.</p>
+<p><em>
+    <strong>FasterXML 2.3.2</strong> <br/><br/>
```
Does this binding really include FasterXML JSON? This looks like a copy&pasta-error :smile:

lewie replied:
OH YES! copy&pasta-error! 
Thanks, I have bookmarked it.

hakan42 commented:
@lewie , thank you for your efforts to get this binding created and added to OH :smile: 

There are still some more minor style issues to be resolved, but I believe those can be best worked on when the binding really sees some widespread usage.

Therefore, merging this PR now, and will ask @lewie for further fixes in seperated PRs.

lewie commented:
@hakan42, 
I 100% agree with you!
It is a good start, nothing is cast in stone.  :smile:

I hope this Bundle will get live and go better and better through widespread usage and several contributers observational/usage perspectives.

The next PR is already on with the new yank version 3.1.0 for java7/8 used over Maven (at the moment I have done this by a snapshot):
````
<dependency>
    <groupId>org.knowm</groupId>
    <artifactId>yank</artifactId>
    <version>3.1.0</version>
</dependency>
````
I have bookmarked.

watou commented:
@lewie, please submit a new PR to tie the new JDBC persistence into the higher-level pom.xml files so it's part of the build and distribution.

hakan42 commented:
@watou , thank you, added the binding to the poms.
